PDF splitting in rest client - api v2

jioffe502 opened the pull request:
## Description

The goal is to create a V2 REST API that preprocesses PDFs by splitting them into single-page subjobs before submission to Redis and Ray, improving processing throughput.

Current progress with pdf splitting without changes to any Pydantic schemas.

The main challenge right now is aligning the V2 job submission with the existing `IngestJobSchema` validation.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/nv-ingest/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] If adjusting docker-compose.yaml environment variables have you ensured those are mimicked in the Helm values.yaml file.

jperez999 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,444 @@
+    operation_id="submit_job_v2",
+)
+async def submit_job_v2(request: Request, response: Response, job_spec: MessageWrapper, ingest_service: INGEST_SERVICE_T):
+    with tracer.start_as_current_span("http-submit-job-v2") as span:
```
can we turn this tracer logic into a decorator. Hopefully it could be applied as a decorator wherever necessary, given the name is actually the name of the function.

jperez999 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,444 @@
+)
+
+
+def split_pdf_to_pages(pdf_content: bytes) -> List[bytes]:
```
can we set this up to use logic, that allows the user to decide the split page interval (i.e. could be 1 but maybe I to break the pdf up into subsets of 10 pages).

jperez999 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,444 @@
+                page_count = get_pdf_page_count(pdf_content)
+                
+                # Split if more than 3 pages
+                if page_count > 3:
```
Why do you need to have this check here? Shouldn't the split logic detect if it needs to split or not?

jperez999 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,444 @@
+    summary="Fetch the result of a previously submitted job by its job_id (V2 with aggregation)",
+    operation_id="fetch_job_v2",
+)
+async def fetch_job_v2(job_id: str, ingest_service: INGEST_SERVICE_T):
```
This is a very long function, and that base if/else statement makes me feel like we can break this up into smaller functions that we can then put together in the fetch_job_v2 function to get desired behavior. Is there a reason this logic is not split into smaller subsets, maybe I am missing something?

jioffe502 replied:
this initially was very defensive programming where I totally keep the fetch Logic for v1 if there is no parent - child mapping in redis. 

im planning on simplifying it down a lot just for now at this point in the draft keeping it verbose

jperez999 reviewed (commented):
Great start, moving in the right direction.

drobison00 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,507 @@
+            end_zero = min(start_zero + pages_per_chunk, total_pages)
+            page_indices = list(range(start_zero, end_zero))
+
+            new_pdf = pdfium.PdfDocument.new()
```
I think when we do this, we're going to lose meta information on all the child PDF's from the parent. Not 100%, but we should probably check.

drobison00 replied:
They could probably be taken from the parent PDF and set on the child.

That said, if this is all we're losing, it may not be necessary.

drobison00 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,507 @@
+            # Add custom attributes to the span
+            span.set_attribute("http.method", request.method)
+            span.set_attribute("http.url", str(request.url))
+            span.add_event("Submitting file for processing (V2)")
```
Re: Julio's comment on monolithic functions. 

We could probably make this a modular helper so its more declarative.

drobison00 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,507 @@
+                        parent_uuid = uuid.uuid4()
+
+                    for chunk in chunks:
+                        chunk_number = chunk["chunk_index"] + 1
```
Re: Juilo's monolithic function comment

This could be its own function to process each chunk, you could also get away with issuing all chunks asynchronously and then wait on all of them at the end of the loop.

drobison00 commented on src/nv_ingest/api/v2/ingest.py:
```diff
@@ -0,0 +1,507 @@
+            subjob_results = []
+            failed_subjobs: List[Dict[str, object]] = []
+
+            for page_index, descriptor in enumerate(ordered_descriptors, 1):
```
This can be made parallel, I would issue all the subjob status requests together then process them as the results become available. `asyncio.gather`

jioffe502 commented:
Added V2 ingest API that pre-splits PDFs at a configurable page threshold, fans out chunked subjobs, and aggregates results before returning.
Updated REST client, service wiring, and docker env to select the new v2 endpoints and expose PDF_SPLIT_PAGE_COUNT.
Documented the workflow and backfilled tests: new dc20_v2_e2e case plus unit coverage for PDF splitting and aggregation helpers.

Additionally, added unit coverage for the new V2 PDF splitting and aggregation helpers, including chunk metadata, subjob preparation, and the submit_job_v2 split flow. Exercised the full parent fetch path—state checks, batched result retrieval, aggregation, and state updates.

DATA: averaged over 8 runs on an H100 NVL

PDF_SPLIT_PAGE_COUNT = 2 -> 50.22s e2e, 9.98 pps (highest stddev)
PDF_SPLIT_PAGE_COUNT = 4 -> 36.77s e2e, 13.5 pps
PDF_SPLIT_PAGE_COUNT = 8 -> 36.05s e2e, 13.77 pps 
PDF_SPLIT_PAGE_COUNT = 16 -> 36.38s e2e, 13.67 pps
PDF_SPLIT_PAGE_COUNT = 32 -> 41.38 e2e, 12.02 pps

Testing Guide:

### Two .env Files

**1. `.env` (root) - Infrastructure**
```bash
PDF_SPLIT_PAGE_COUNT=32    # Pages per chunk

NV_INGEST_MAX_UTIL=32
MAX_INGEST_PROCESS_WORKERS=32
INGEST_DISABLE_DYNAMIC_SCALING=true

# And the most recent NIM RCs 
```

**2. `scripts/tests/.env` - Test settings** 

Refer to nv-ingest/scripts/tests/README.md for more information on how to set this up

```bash
DATASET_DIR=/raid/jioffe/bo20
TEST_NAME=dc20_v2 # this will be what populates your artifacts directory
```

---

### Test V2

```bash
# Run the v2 test (AUTOMATICALLY uses v2 API):
python scripts/tests/run.py --case=dc20_v2_e2e --doc-analysis

# Expected output for bo20:
# ✓ V2 results match V1 expected values!
# text=496, tables=164, charts=184
```

---

### Test Different Split Sizes

```bash
# Edit ./.env → change PDF_SPLIT_PAGE_COUNT=8

# Restart docker services
docker compose down
docker compose --profile retrieval --profile table-structure up -d

# Run tests again
python scripts/tests/run.py --case=dc20_v2_e2e --doc-analysis
```

### Key Points

- **Root `.env`**: Controls PDF splitting (infrastructure setting)
- **`scripts/tests/.env`**: Controls test parameters (dataset, hostname, etc.)
- **V2 auto-enabled**: `dc20_v2_e2e.py` sets `NV_INGEST_API_VERSION=v2` automatically
- **Results**: Should match V1 exactly

jperez999 commented on scripts/tests/cases/dc20_v2_e2e.py:
```diff
@@ -0,0 +1,234 @@
+    return f"{test_name}_{_now_timestr()}"
+
+
+def main() -> int:
```
This test case does not have any asserts. Is there a situation where this could fail silently but report as a success?

jioffe502 replied:
@jperez999 dc20_v2_e2e is our e2e integration test

jperez999 replied:
right it is an integration test but are we checking for the results to ensure we are actually getting back what is exected? Do the other integration tests follow this same strategy? Seems here like you are returning 0 regardless of whether the results are the same or not. Shouldnt this fail if the results are different? IMO if you changed the if statement at the end to an assert you would be better off because it would fail if that assert failed.
